### PR TITLE
Generalize global constraint variables to `module V: Printable.S` in each analysis

### LINKED ISSUE #469: Generalize global constraint variables from `varinfo`
Each activated analysis should be able to define its own global constraint variables (and their type). All these should correspond to different constraint variables in the constraint system such that their dependencies are more fine-grained and independent.

Keeping them separate could also improve precision because widening happens separately per-analysis.

### PR DESCRIPTION
**Pull request by sim642**
Closes #469.

### Changes
1. Adds `module V: Printable.S` to each analysis spec, which will be the module used for global constraint variables for that analysis. Since it's abstract `V.t` can be anything, not necessarily `varinfo`.
2. Adds `'v` type parameter to `ctx`. This is necessary to make `ctx.global` and `ctx.sideg` correctly typed for that `V.t`.
3. Rewrite `MCP`'s globals handling to use "variants" instead of lists for both global variables and their values. Each variant is just a pair of the analysis ID (`int`) and the corresponding variable/value as `Obj.t`.
4. Refactor some analyses to use non-`varinfo` globals: thread, threadJoins, arinc and extract_arinc.
5. Require analyses to explicitly define `V` if they want to use any globals (including `varinfo`). `DefaultSpec` just uses an empty (variant) type, which means no globals can be used at all.
6. Fix g2html to handle the changed XML where globals have just one analysis value, not all.

### TODO
- [x] Comparison runs on sv-benchmarks to see if the abstract variant for globals has any overhead compared to straight `varinfo`. Also to see if immediately any performance or precision improvement arises from the separation of globals for different analyses.
- [x] Refactor region analysis as well, after removing the dependent shapes analysis (#474, #460).
- [x] Fix Gobview build.
- [ ] Refactor access analysis in #391 to avoid grouping all struct-based accesses together.

### CODE REVIEW

**Comment by michael-schwarz** [comment]
Looks good thus far! I have to say I am thoroughly impressed that this did not require a major overhaul of everything!

**Comment by michael-schwarz** [comment]
What would be interesting is if the number of globals that are widening points changed, the hypothesis being that less widening might be applied. But then this is super hard to measure because we now have a lot more globals than we did before.

Also, it would be interesting to see what happens with runtime and the number of variables and evaluations if one runs one of the larger multi-threaded programs (e.g. the traces or more traces bechmarks), or some of the other larger projects from our test repo with `earlyglobs`.

**Comment by sim642** [comment]
> Also, it would be interesting to see what happens with runtime and the number of variables and evaluations if one runs one of the larger multi-threaded programs (e.g. the traces or more traces bechmarks), or some of the other larger projects from our test repo with `earlyglobs`.

I ran the traces benchmarks before and after, didn't notice much of a difference there either (and the bigger programs were from sv-benchmarks anyway). I don't know about the single-threaded benchmarks in bench though.

It's also the case that the base analysis and privatizations still use `varinfo` variables for everything. I don't know yet if there are any real opportunities to use the new generalization there, because most of the additional global variable components (background lockset, write lockset, thread ID) still need to be universally quantified over, so the (nested) maps need to remain to make that possible.

**Comment by sim642** [comment]
> It's also the case that the base analysis and privatizations still use `varinfo` variables for everything. I don't know yet if there are any real opportunities to use the new generalization there, because most of the additional global variable components (background lockset, write lockset, thread ID) still need to be universally quantified over, so the (nested) maps need to remain to make that possible.

I have now also refactored all the privatizations at the top level at least. So no more `MUTEX_GLOBAL_g` and `MUTEX_INITS` varinfos any more. Similarly for base protection-based I separated the unprotected and protected globals. Moreover, I replaced the global product domains for those things with `Lift2`s.

**Comment by sim642** [comment]
I ran sv-benchmarks again after my futher refactorings and there is now a 3% improvement in the regression line compared to `master`:
image
(Full comparisons here: https://goblint.cs.ut.ee/results/50-all-fast-gen-glob-after-refactor/results/).

I'm not sure where this improvement exactly came from though because the changes were to privatizations, which only matter for concurrent benchmarks and even then we don't use any of the more expensive privatizations.

File: src/analyses/basePriv.ml [hunk 1]
```diff
@@ -293,19 +294,10 @@ struct
...
-    | `Return -> (* required for thread return *)
-      begin match ThreadId.get_current ask with
-        | `Lifted x when CPA.mem (ThreadIdDomain.Thread.to_varinfo x) st.cpa ->
-          let x = ThreadIdDomain.Thread.to_varinfo x in
-          let v = CPA.find x st.cpa in
-          sideg (mutex_global x) (CPA.singleton x v);
-          {st with cpa = CPA.remove x st.cpa}
-        | _ ->
-          st
-      end
```
Comment (michael-schwarz) [inline src/analyses/basePriv.ml:308]: Why can we now remove this?

**Review by michael-schwarz — APPROVED** [review]

**Pull request merged**